### ISSUE TITLE
Solid Modularity

### ISSUE THREAD

**Issue by sandhawke** [solid/solid#72]
I think it would be easier to talk about solid in a wider community and get interoperability if solid were defined as a collection of specs.     There should be no solid-spec itself, except as a simple list.  Also, I think we need release numbering.   (with semver.)

So maybe "Solid Baseline, release 1.3" is:
* https head
* https get
* https put
* https post
* https patch
* ldp basic containers
* ldp rel=describedby
* webid-tls
* wac 
* posix vocab on containers
* globbing
* account creation 
* watch-via-websockets
* something about resource discovery
* something about finding the webid from the site url
* something about https options
* something with patch using sparql update
* con-neg between turtle and json-ld

This is (approximately) a list of things you need to understand to implement a solid 1.3 server and list of things available to you, if you want them, in writing a solid 1.3 application.

It would be an interesting exercise to retroactively figure out a version number based on the compatible and incompatible changes made since the first version of ldphp.

**Comment by melvincarvalho** [comment]
:+1: in general, from a technical POV this helps.

From a branding perspective there may be other considerations (I'll try to stay out of the branding discussion and let wiser people than me decide).

**Comment by melvincarvalho** [comment]
Question.  Is there a dependency tree here?  

For example does HTTP GET imply HTTP HEAD?  

Or does WebID TLS imply WebID.

**Comment by sandhawke** [comment]
Yes, there's a dependency tree.    That's why I didn't list Turtle or RDF.

But I don't think the spec for HTTP GET says you have to implement HTTP HEAD or visa versa, so I list them separately.

Ideally for each item we'd have a section of the test suite to help illustrate and debug this.  But it wouldn't try to be exhaustive I think -- we wouldn't include the whole turtle test suite as part of the rel=describedby inclusion -- but we would do a few turtle-ish things.

**Comment by msporny** [comment]
+1 to the proposal to break solid up into different specs. 

We've been thinking about implementing bits and pieces of SoLiD in our Bedrock application server: https://digitalbazaar.gitbooks.io/bedrock/content/introduction.html

Unfortunately, at the moment we don't have a business reason to do so, so it's been relegated to "weekend fun projects" for folks at the company. Every time we've looked at SoLiD, there are pieces that we like and would like to re-use but have a hard time figuring out if we can implement just a small piece of SoLiD.

To be clear, don't do anything on our account. Even if you break SoLiD up into different specs, it'll be a while before we can find a business justification for implementing it unless there is a clear and present problem that it solves for us, and thus far, we're not quite in the social web space - more education, government, defense, financial services, and healthcare for now.

These are the bits that look interesting to us so far:

* wac
* posix vocab on containers
* globbing
* account creation
* watch-via-websockets
*  resource discovery
* finding webid from the site url
* something with patch using sparql update (but not necessarily the SPARQL bits - we'd be interesetd in linked data patch w/o SPARQL)

**Comment by sandhawke** [comment]
I wonder what to call these things and how to make then easier to specify nicely.    Web service modules.  Web protocols.    LDP components.   Restful API Patterns. ... ?

**Comment by sandhawke** [comment]
Probably a good principal is that all modules are detectable.    If they're not there, in the form you expect, the client gets an error and has a chance to try to do it differently.   This could be hidden by a library, like a polyfill.

**Comment by akuckartz** [comment]
:+1: for the modular approach

**Comment by bblfish** [comment]
+1 for being modular. Indeed I suggested that  in (issue 36 of solid-spec: split SoLiD spec into mini specs)[https://github.com/solid/solid-spec/issues/36].

But I am not sure it is worth specifying things out at the level of `HTTP` verbs, as that is already done by LDP. So unless people here feel the need to re-do the work of that group - which took 3 years - it may be better to split SoLiD along different features that are separable. And as far as possible to use specifications that exist, where SoLiD just needs to provide integration documentation.

**Comment by sandhawke** [comment]
Thinking about this a little more.

Imagine Solid v12.6 has a certain set of modules, A, B, and C.   And something's come along that's really better than A, called A2.    While Solid is still at 12.6, servers can start offering A2 in parallel.   Clients can start looking for A2 and using it instead of A if they find it.   This involves no change to the baseline.    Thing are good, except client developers are wondering if they can start to rely on A2.

So maybe in v12.7 we include A2 (along with A).  To claim 12.7 compliance, a server will have to implement A2.     Hopefully everyone does this...   

Now A starts to look pretty lame and annoying to support.   People want to remove it.  So v13 comes out with A removed, and as servers upgrade, ... any software still using A breaks.   So, some servers keep A in there, even though it's not required in v13.

That's ... not great,    Not sure what's better.    Lots of things kind of like this out there.  I don't feel like anyone's really solved it.   It's rare that things get removed.

Probably having something like caniuse.com is important, with the available solid servers cross-referenced

**Comment by sandhawke** [comment]
Thanks for the link @bblfish 

I agree splitting HEAD and GET is a little pedantic.   But surely PATCH needs to be separate.   And PUT is kind of the core of solid -- the fact that solid servers actually implement PUT, unlike most things that claim to implement HTTP.

The work already done by LDP-WG doesn't need to be considered wasted, it was just part of the road to getting us here.   It was already clear to some of us in the LDP-WG that this kind of modularity was needed, but we ran out of steam before making it happen.  See eg  https://lists.w3.org/Archives/Public/public-ldp-wg/2014Nov/0051.html and https://www.w3.org/2012/ldp/wiki/LDPNext

**Comment by bblfish** [comment]
yes, just trying to see that we minimise the work. It's true that a future version of LDP should also be more modularised as you suggested in the last month of the LDP group. I think we should in any case pretend that LDP is 3 specs, one for each of the containers.

On that topic, we also need to consider `SEARCH` as well as `PATCH` , and yes specify new containers that make `PUT` useable.

**Comment by sandhawke** [comment]
That was 8.5 months before the end of LDP

**Comment by kidehen** [comment]
+1 to modularity such that interop can also be much modularized. For instance we have key interop points regards:

[1] LDP -- can all implementations perform basic operations across one another using this protocol?
[2] WebID+TLS -- can clients independently authenticate against IDPs that implement this protocol? 
[3] WebACL -- can basic ACLs be tested across any HTTP client and servers that implement WebACL (which depends on #2, at the very least). 

What we shouldn't do is keep on adding new subtleties under new headers and then magically expecting any interop to happen because it won't, as history has shown.

**Comment by sandhawke** [comment]
@kidehen can you give me a concrete example of what you're w saying we shouldn't do?    I'm not quite understanding you.

**Comment by nicola** [comment]
+1 for modularity, this aligns perfectly my original idea of writing the different parts in different modules, so that users like @msporny can pick and mix (also, great that bedrock is in js!)

**Comment by melvincarvalho** [comment]
@nicola good point.  I wonder if there might also be parts of bedrock that could be reused in Solid?

One interesting section I noticed:

"A bedrock module that adds website or REST API authentication to bedrock via passport."

https://digitalbazaar.gitbooks.io/bedrock/content/bedrock-passport.html

**Comment by msporny** [comment]
> I wonder if there might also be parts of bedrock that could be reused in Solid

More than likely, not - Bedrock modules work in the bedrock application framework and have dependencies. There are 43+ bedrock modules covering things like MongoDB and Redis support, account management, public/private key management, Linked Data Signatures, mail, job scheduling, etc.

Also note that Bedrock isn't open source in the traditional sense. It's on Github and we do provide a free-for-use non-commercial license in certain student/R&D/hobbyist scenarios, but anything further than that requires a commercial license. This is one of the ways our company makes money (if the people using Bedrock are making money). Bedrock funds a good chunk of our "donated" standards work at W3C, IETF, and other fora... just making that abundantly clear.

More info on the Bedrock non-commercial license here:
https://github.com/digitalbazaar/bedrock/blob/master/LICENSE.md

**Comment by kidehen** [comment]
What we shouldn't do is start a new focal point of compliance and compatibility with "SoLID" as the focal point since its a collection of many things. Let's actually establish what we are doing in regards to:

[1] Identifiers -- WebID
[2] Identity Claims -- WebID-Profile Document
[3] Identity Claims Verification -- WebID+TLS Protocol
[4] Resource Access Controls -- WebACL
[5] Read-Write Operations against Resources -- LDP .

Being able to collectively demonstrate interop based on 1-5 would be a major milestone e.g.:

[1] Successful login across compliant services
[2] Creating resources on compliant servers
[3] Providing controlled access to WebACL controlled resources on compliant servers.

If this isn't clear enough, I really don't have the bandwidth to add more since we've already jumped through the hoops are are ready to perform interop tests with anyone that's also invested time in implementing 1-5 using existing open standards.

**Comment by melvincarvalho** [comment]
@kidehen thanks for creating these implementations.  It is appreciated you have limited bandwidth, so why dont we (over another channel) try and put together a testing plan, broken down into milestones for parts 1-5 (for any team that wants to try to interop)?  

From my experience, finding the time to set up tests an debug is a challenge.  What kind of timeframe do you think could be feasible for this -- my own intuition is 1-2 points (from 5) per quarter is achievable, leading to 1-5 interop by end of the year -- is that too optimistic, or not optimistic enough?  

Thoughts?

**Comment by sandhawke** [comment]
The thing is, @kidehen, the specs you enumerate aren't a perfect fit for what we're trying to do here (namely end software lock-in, especially social lock-in).   They are steps in the right direction, yes, but they are neither necessary nor sufficient.

**Comment by kidehen** [comment]
@sandhawke : I really don't understand how you can dismiss my fundamental point when as I type, our tools (courtesy of the open standards listed) already moves data across application silos with ease. I don't know how you perceive data de-silo-fication, but I know that it boils down to structured data representation via an abstract language (e.g. RDF), notations (Microdata, JSONL-LD, Turtle), serialization formats (Microdata, JSONL-LD, Turtle), and a transmission protocol (e.g, HTTP). Now, if you know something better, that you can demonstrate, I am all ears. At this point I've implemented my claims and they work very well. All I am doing is using what exists.

**Comment by kidehen** [comment]
@melvincarvalho -- If we can't use the file create, save, and share pattern as the basis of basic interop with regards to document creation and sharing, using existing open standards from the existing collection of W3C standards, I really do not know what to say. What I don't have time or patience for is endless reconstruction of specs without actual implementations tested for interoperability. 

Whenever anyone is ready, I'll test our stuff against it. I've already done so against most of your apps with success regarding login related authentication using WebID, WebID-TLS and WebID-Profile.

I think you need to have others get to this point, against your collection of apps. I am there and beyond.

**Comment by sandhawke** [comment]
@kidehen application silos (where two applications can run on the same machine, under user control, but still not talk to each other) are the old bane of interoperability, and I can easily imagine your tools work well on that front.   The goal of our crosscloud projects (of which solid is a part) is to extend that to social lockin, where you're not just locked in because of where your data is, you're locked in because that's where your contacts are.    The crosscloud idea -- force social apps to interact with user data across the network, so the data can remain under user control -- gives us a general direction.   RDF and RWW and WebID-TLS and WAC give us a very rough first cut.   Cimba showed you could build something like twitter with a crosscloud architecture.   But Cimba wasn't actually usable, in practice, for a number of reasons, which we're gradually making progress toward solving.    But we don't know exactly how to do it.   It's research.   So, at the moment, I can't just give you a spec to implement.

**Comment by melvincarvalho** [comment]
@sandhawke solid is a spec, it can be implemented.  Maybe crosscloud isnt.  But this thread is about implementing Solid.  Part of that should be a way testing components across different systems.

@kidehen propose a time and communication channel for interop testing, lets try and clear at least one hour, maybe a doodle poll?

**Comment by sandhawke** [comment]
@melvincarvalho I don't think that comment is appropriate in this context.   Please rephrase or remove it.

**Comment by kidehen** [comment]
@sandhawke : I repeat, our tools are all about data de-silo-cation . We can break down any data silo using existing W3C standards and/or proprietary protocols. Silos exists because social networks and apps control document storage and document content types. When you get a moment digest all the nanotation based data de-silo-fication demos I've been sharing. To make matters easier install a copy of our OSDS browser extension. 

Here's our fundamental difference, I believe key Web standards from the W3C  are in place for social network players to adopt, whenever they choose. I don't believe in an endless quest to direct social networking players to adopt another bag of new and unproven de facto standards specs. History shows the aforementioned approach will never work. Nobody spec'd the Web, it happened and folks adopted it (at varying levels) to reduce "opportunity cost". 

Tools that utilize standards win 100% of the time. I never seen a standard that's lead to broad adoption, just because of the existence of its spec. 

@melvincarvalho : I'll ping you if I hit an interop issues, as I usually do. That's productive use of my time. 

Links:

[1] http://osds.openlinksw.com 
[2] https://medium.com/@kidehen/openlink-structured-data-sniffer-osds-2-6-1-unleashed-15137e558350#.ketiymmjh -- post with live examples re. nanotation.

**Comment by sandhawke** [comment]
Thinking about it a little more, I think we have both:

* Pod Behaviors (eg allowing PUT to create a resource)
* App Behaviors (eg when to use PUT to create a resource)

For example, Issue #74 looks like it would be resolved by an AppBehavior spec.  Meanwhile, #61 is a PodBehavior concern.

**Comment by dmitrizagidulin** [comment]
Solid spec is now broken into modules as of PR https://github.com/solid/solid-spec/pull/87. Closing this issue.
